### PR TITLE
fixes parsing of 'this' in arrow iife

### PR DESCRIPTION
**Pull request by sokra**
yarnpkg/berry#1832

**What kind of change does this PR introduce?**
bugfix

**Did you add tests for your changes?**
yes

**Does this PR introduce a breaking change?**
no

**What needs to be documented once your changes are merged?**
nothing

### CODE REVIEW

File: test/cases/parsing/this-in-arrow-iife/index.js [hunk 1]
```diff
@@ -0,0 +1,14 @@
+function x() {
+	let value;
+	(() => {
+		value = this;
+	})();
+	return value;
+}
+
+it("should parse this in an IIAFE correctly", () => {
```
Comment (vankop) [inline test/cases/parsing/this-in-arrow-iife/index.js:9]:
```suggestion
it("should parse this in an IIFE correctly", () => {
```
as I understand..

File: lib/javascript/JavascriptParser.js [hunk 2]
```diff
@@ -2535,7 +2535,7 @@ class JavascriptParser extends Parser {
 		const renameThis = currentThis ? getVarInfo(currentThis) : null;
 		const varInfoForArgs = options.map(getVarInfo);
 		const wasTopLevel = this.scope.topLevelScope;
-		this.scope.topLevelScope = arrow ? "arrow" : false;
+		this.scope.topLevelScope = wasTopLevel && arrow ? "arrow" : false;
```
Comment (vankop) [inline lib/javascript/JavascriptParser.js:2538]: maybe enum make sense

**Review by vankop — APPROVED** [review]

File: lib/javascript/JavascriptParser.js (hunk 2 above)
Comment (sokra) [reply lib/javascript/JavascriptParser.js:2538]: Isn't `true | false | "arrow"` an enum?

Comment (vankop) [reply lib/javascript/JavascriptParser.js:2538]: Technically this is [union type](https://www.typescriptlang.org/docs/handbook/unions-and-intersections.html#union-types), anyway this is just design question.. In my opinion such design easier to maintain (you don't need to expose type), if you understand what `"arrow"` does. On the other hand enum allows to write self-described code -  e.g. `Scope.ARROW_FUNCTION` is more understandable. Maybe this is not worth it in current case..

Comment (sokra) [reply lib/javascript/JavascriptParser.js:2538]: Definitely makes sense, but this was chosen for backward compat with webpack 4, which had `true | false`. As `"arrow"` is truthy this allows to keep most existing code.

**Pull request merged**